mh/pat.py

```diff
@@ -84,14 +84,17 @@ def layer_broadcast(self, session, packet_id, data, seq,
 
     def circle_broadcast(self, circle, packet_id, data, seq,
                          session=None):
-        with circle.lock():
+        handlers = []
+        with circle.lock(), circle.players.lock():
             for _, player in circle.players:
                 if session and player == session:
                     continue
 
                 handler = self.get_pat_handler(player)
                 if handler:
-                    handler.try_send_packet(packet_id, data, seq)
+                    handlers.append(handler)
+        for handler in handlers:
```

Comment:
I suppose you do this dance of collecting the handlers first and then send the payload too avoid a deadlock?

Reply:
That's part of it. It's mostly that it's bad practice to dispatch packets (or do any unnecessarily lengthy operation) with a lock in place- it slows down the program. Better to handle the packet dispatching once the lock is terminated.

I have a separate branch (based off this one, so no PR yet) in which I generalized this concept into a PacketQueue class that is entered using a `with` statement alongside locks that have packet dispatchers inside them. That will enforce this "no packet sending until the lock is done" rule throughout the server.